### PR TITLE
8333358: java/io/IO/IO.java test fails intermittently

### PR DESCRIPTION
**Pull request by pavelrappo**
Please review this fix for an intermittent test failure.

On some configurations, the default `expect` timeout of 10 seconds is insufficient. It is increased to 20; it's hard to imagine a configuration for which that new value would still be insufficient, but we'll see.

Aside from that, test-generated diagnostics are improved: the version of the `expect` command and the duration of each test method are recorded. `output.exp` is modified for robustness and clear indication of the timeout condition.

I was reminded, out-of-band, that test timeouts should be scaled with `test.timeout.factor`. I propose to integrate this PR first and then separately update all the affected tests to scale their `expect` timeouts.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333358](https://bugs.openjdk.org/browse/JDK-8333358): java/io/IO/IO.java test fails intermittently (**Bug** - P3)

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19627/head:pull/19627` \
`$ git checkout pull/19627`

Update a local copy of the PR: \
`$ git checkout pull/19627` \
`$ git pull https://git.openjdk.org/jdk.git pull/19627/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19627`

View PR using the GUI difftool: \
`$ git pr show -t 19627`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19627.diff">https://git.openjdk.org/jdk/pull/19627.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19627#issuecomment-2158227169)

### CODE REVIEW

File: test/jdk/java/io/IO/IO.java [hunk 1]
```diff
@@ -174,4 +185,40 @@ public void nullConsole(String method) throws Exception {
         assertEquals(1, output.getExitValue());
         output.shouldContain("Exception in thread \"main\" java.io.IOError");
     }
+
+
+    // adapted from https://raw.githubusercontent.com/junit-team/junit5/main/documentation/src/test/java/example/timing/TimingExtension.java
```
Comment (naotoj) [inline test/jdk/java/io/IO/IO.java:190]: Probably better to refer to the user guide on the web (https://junit.org/junit5/docs/snapshot/user-guide/#extensions-lifecycle-callbacks-before-after-execution)

**Review by naotoj — APPROVED** [review]
LGTM

**Comment by pavelrappo** [comment]
Preliminary testing on some affected configurations showed that even 20 seconds might not be enough. Initially, I suggested that it would be hard to imagine, yet here we are.

A good solution would be to disable `expect` timeout and, thus, fall back onto jtreg timeout. The default jtreg timeout is 120 seconds, and `IO.java` does not override it. Jtreg timeout can be adjusted by the `-timeout` command-line option.

I believe, this would be the most flexible solution for this timeout issue.

File: test/jdk/java/io/IO/IO.java [hunk 2]
```diff
@@ -174,4 +185,40 @@ public void nullConsole(String method) throws Exception {
         assertEquals(1, output.getExitValue());
         output.shouldContain("Exception in thread \"main\" java.io.IOError");
     }
+
+
+    // adapted from https://junit.org/junit5/docs/current/user-guide/#extensions-lifecycle-callbacks-timing-extension
+
+    public static class TimingExtension implements BeforeTestExecutionCallback,
```
Comment (AlanBateman) [inline test/jdk/java/io/IO/IO.java:192]: In passing, this feels like infrastructure that we could use in several areas as it regularly comes up as to how long specific tests take (the "seconds" line recorded by jtreg in the .jtr file is too coarse grain when there are many tests in the same source file).

**Comment by jaikiran** [comment]
> Preliminary testing on some affected configurations showed that even 20 seconds might not be enough. Initially, I suggested that it would be hard to imagine, yet here we are.

I haven't seen the jtreg launch command or the jtr files for these timeout failures, but in the past on systems that Matthias has run into similar issues, it was observed that those systems run with a very high `-concurrency` value (which implies too many test processes running at the same time) which could cause slowness. I don't know if that's what is a contributing factor here. 

> A good solution would be to disable expect timeout and, thus, fall back onto jtreg timeout. 

I think that's a better solution. That removes the part where we have to guess what's the best/appropriate timeout value to use in the test.

File: test/jdk/java/io/IO/IO.java (hunk 2 above)
Comment (jaikiran) [reply test/jdk/java/io/IO/IO.java:192]: Looking briefly at this, I think we should be able to provide this as a feature of jtreg's support of `junit` test itself. I've created https://bugs.openjdk.org/browse/CODETOOLS-7903751 to investigate this support in jtreg.

Comment (AlanBateman) [reply test/jdk/java/io/IO/IO.java:192]: This is something that I discussed with Christen Stein about. We've tossed around some ideas, he created CODETOOLS-7903752 so we may have a dup.

Comment (jaikiran) [reply test/jdk/java/io/IO/IO.java:192]: I've closed mine as the duplicate of the other.

**Comment by pavelrappo** [comment]
> in the past on systems that Matthias has run into similar issues, it was observed that those systems run with a very high `-concurrency` value (which implies too many test processes running at the same time) which could cause slowness

In jtr files from those systems, I didn't see any `-concurrency` values, which I believe are not recorded there, but I saw `test.timeout.factor=9.0`, which is by itself is very telling.

**Review by naotoj — APPROVED** [review]

**Pull request closed without merging**